File: ai_research/dataset_generation/domain/llm_response.go
package domain

import (
	"log"
	"regexp"
	"strings"
)

func ParseExamples(response string) []Example {
	exampleSections := parseExampleSections(response)
	examples := make([]Example, 0, len(exampleSections))
	for _, section := range exampleSections {
		example, err := parseExample(section)
		if err != nil {
			log.Printf("Failed to parse example, skipping: %v", err)
			continue
		}
		examples = append(examples, example)
	}
	return examples
}

func parseExampleSections(response string) []string {
	re := regexp.MustCompile(`### .*Example.*\n\s*`)

	// Find all headers
	matches := re.FindAllStringSubmatchIndex(response, -1)

	sections := make([]string, len(matches))

	// Extract content between headers
	for i := 0; i < len(matches)-1; i++ {
		start := matches[i][1] // End of the current header
		end := matches[i+1][0] // Start of the next header
		section := strings.TrimSpace(response[start:end])
		sections[i] = section
	}

	// Extract content after the last header
	if len(matches) > 0 {
		start := matches[len(matches)-1][1] // End of the last header
		section := strings.TrimSpace(response[start:])
		sections[len(matches)-1] = section
	}

	return sections
}

File: ai_research/dataset_generation/domain/llm_response.py
import logging

from ai_research.dataset_generation.domain.example import Example, parse_example

logger = logging.getLogger(__name__)


def parse_example_sections(response: str) -> list[str]:
    sections: list[str] = []
    current_section: list[str] | None = None
    in_fence = False

    for line in response.splitlines(keepends=True):
        if line.startswith("```"):
            in_fence = not in_fence

        is_example_heading = (
            not in_fence and line.startswith("### ") and "example" in line.casefold()
        )
        if is_example_heading:
            if current_section:
                sections.append("".join(current_section).strip())
            current_section = []
            continue
        if current_section is not None:
            current_section.append(line)

    if current_section:
        sections.append("".join(current_section).strip())
    return sections


def parse_examples(response: str) -> list[Example]:
    examples: list[Example] = []
    for section in parse_example_sections(response):
        try:
            examples.append(parse_example(section))
        except ValueError as exc:
            logger.warning("Failed to parse example, skipping: %s", exc)
    return examples
